[MRG] Fix predict_proba not fitted check in SGDClassifier

aniruddhadave opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #10938 


#### What does this implement/fix? Explain your changes.
Fixes not fitted check in predict_proba method so that it doesn't throw a not fitted error while referencing the method. Checks whether the classifier is fitted or not when the method is called.


#### Any other comments?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

aniruddhadave committed:
Fix Issue 10938
- Remove not fitted check from predict_proba method of SGDClassifier
- Check only while calling predic_proba

jnothman reviewed (commented):
I think that is the right fix. Please add a non-leash regression test

aniruddhadave commented:
I understand non-regression testing as given on the contributing guidelines but how is a non-leash regression test different from that?

lesteve commented:
I think @jnothman meant a non-regression test (maybe autocorrect or something?). Great if you understand what it is from the contributing guidelines. Can you add a non-regression test then?

lesteve commented:
Suggestion for non-regression test (add a function insklearn/linear_model/tests/test_sgd.py):
```py
from sklearn.linear_model import SGDClassifier

clf = SGDClassifier()
clf.predict_proba
clf.predict_log_proba
```

aniruddhadave commented:
@lesteve There already exists a test case for predict_proba method (test_sgd_proba) shouldn't it be modified instead of writing a separate test case?

lesteve commented:
Not really important, but I would rather put that in a separate test function. It is a bit a special case to just test that you can access the `predict_proba` and `predict_log_proba` attribute.

aniruddhadave committed:
Add test case for issue #10938
	-Test if the the predict_proba and predict_log_proba
	 methods can be accessed before fitting
	-Test if the not fitted check is performed before calling
	 the proba methods

jnothman commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -467,6 +468,26 @@ def test_set_coef_multiclass(self):
+
+            # Checks if not fitted check is performed while calling 
+            # the methods
+            assert_raises(NotFittedError, clf.predict_proba,[[3,2]])
```
Space after commas please

jnothman commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -467,6 +468,26 @@ def test_set_coef_multiclass(self):
+        # is accessible for refrencing before fitting 
+        # the SGD classifier
+        clf = SGDClassifier()
+        assert_false(hasattr(clf,"predict_proba"))
```
Please just use bare assert not ... rather than assert_false. Thanks

jnothman reviewed (approved):
Otherwise LGTM

aniruddhadave committed:
Issue #10938: Code Cleanup
	-Remove extra line
	-Add space after comma
	-Use assert instead of asser_false

aniruddhadave committed:
Remove Trailing whitespaces

lesteve commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -467,6 +468,26 @@ def test_set_coef_multiclass(self):
+
+            # Checks if not fitted check is performed while calling
+            # the methods
+            assert_raises(NotFittedError, clf.predict_proba, [[3, 2]])
```
I would expect the NotFittedError case to be already tested in test_common.py so I would remove this from the test.

lesteve commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -467,6 +468,26 @@ def test_set_coef_multiclass(self):
+
+        for loss in ["log", "modified_huber"]:
+            clf = SGDClassifier(loss=loss)
+            assert_true(hasattr(clf, "predict_proba"))
```
Use bare assert here too.

lesteve commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -467,6 +468,26 @@ def test_set_coef_multiclass(self):
+        # is accessible for refrencing before fitting
+        # the SGD classifier
+        clf = SGDClassifier()
+        assert not(hasattr(clf, "predict_proba"))
```
Maybe check this for all the losses that do not support predict_proba, you can use `SGD.loss_functions` to get all the possible losses I think.

Slight improvement would be to also check the error message:
```py
with pytest.raises(AttributeError, 'probability estimates are not available for loss={!r}'.format(loss)
```

lesteve reviewed (commented):
Some comments.

aniruddhadave committed:
Add test to check error message
Remove test for NotFittedError

lesteve committed:
Tweaks

lesteve commented:
I pushed some minor tweaks, I think this can be merged when CIs are green.

jnothman commented:
Perhaps this needs a changelog entry. 
Please add an entry to the change log at `doc/whats_new/v0.20.rst` under API changes. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

lesteve commented:
> Perhaps this needs a changelog entry.

I am a bit undecided but I would say that this is rather an obscure problem (trying to only access `predict_proba` rather than calling it before fit) and I feel this does not really deserve a log entry. I am going to merge this one.
